source/Halibut/Transport/ConnectionManagerAsync.cs

```diff
@@ -0,0 +1,300 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Halibut.Diagnostics;
+using Halibut.Transport.Protocol;
+using Nito.AsyncEx;
+
+namespace Halibut.Transport
+{
+    public class ConnectionManagerAsync : IConnectionManager
+    {
+        readonly ConnectionPoolAsync<ServiceEndPoint, IConnection> pool = new();
+        readonly Dictionary<ServiceEndPoint, HashSet<IConnection>> activeConnections = new();
+
+        // We have separate locks for connections in general (including the pool) vs specifically activeConnections.
+        // This is because disposing calls OnConnectionDisposed, which causes deadlocks if we are not careful.
+        readonly SemaphoreSlim connectionsLock = new(1, 1);
+        readonly SemaphoreSlim activeConnectionsLock = new(1, 1);
+        
+
+        public bool IsDisposed { get; private set; }
+
+        public void Dispose()
+        {
+            pool.Dispose();
+            using (connectionsLock.Lock())
+            {
+                IConnection[] connectionsToDispose;
+                using (activeConnectionsLock.Lock())
+                {
+                    connectionsToDispose = activeConnections.SelectMany(kv => kv.Value).ToArray();
+                }
+
+                foreach (var connection in connectionsToDispose)
+                {
+                    SafelyDisposeConnection(connection, null);
+                }
+            }
+
+            IsDisposed = true;
+        }
+
+        [Obsolete]
+        public IConnection AcquireConnection(ExchangeProtocolBuilder exchangeProtocolBuilder, IConnectionFactory connectionFactory, ServiceEndPoint serviceEndpoint, ILog log, CancellationToken cancellationToken)
+        {
+            Tuple<IConnection, Action> openableConnection;
```

Comment:
Should the sync paths just throw? I guess perhaps it is safer if they work at least to some extent.

The tests do pass if these were not implemented right?

Reply:
Like you said, I kept it because of the gradual transition to async.

The connection manager tests pass with those Obsolete methods throwing exceptions.

Until we are fully cutover to async, I'd be nervous making sync methods throw.

I've [created a card](https://app.shortcut.com/octopusdeploy/story/56885/make-sync-methods-in-connectionmanagerasync-and-connectionpoolasync-throw-exceptions) to come back to this.